truss/templates/server/model_wrapper.py

```diff
@@ -68,6 +68,7 @@ class MethodName(str, enum.Enum):
     "starlette.responses.Response",
     pydantic.BaseModel,
 ]
+ModelFn = Callable[..., Union[OutputType, Awaitable[OutputType]]]
```

Comment:
From this [comment](https://github.com/basetenlabs/truss/pull/1364#discussion_r1942125608), we can introduce a slight type constraint, but I'm not sure how much benefit it brings:
- Seems non trivial to have discriminated union functionality to tie the `ArgConfig` and the underlying `Callable` together (i.e. if `ArgConfig.NONE`, then the type system knows it's a `Callable[[], OutputType]`)
- Similar to above, tricky to have the same discrimination on async vs sync

To avoid all that, I made it have variable arguments for now, given that `ArgConfig` will do validation on parse. We have to do a couple explicit casts to make the type system happy as a result though.